File: plugins/plugin-scheduling/typescript/src/providers/scheduling-context.ts
/**
 * @module scheduling-context
 * @description Provider that gives the agent context about scheduling state
 */

import type { IAgentRuntime, Memory, Provider, State } from "@elizaos/core";
import type { SchedulingService } from "../services/scheduling-service.js";
import type { Meeting } from "../types.js";

/**
 * Format a meeting for display in context
 */
const formatMeetingForContext = (meeting: Meeting, service: SchedulingService): string => {
  const timeStr = service.formatSlot(meeting.slot);
  const participants = meeting.participants.map((p) => p.name).join(", ");
  const locationStr =
    meeting.location.type === "in_person"
      ? `at ${meeting.location.name}`
      : meeting.location.type === "virtual"
        ? "virtual meeting"
        : "phone call";

  return `- "${meeting.title}" on ${timeStr} (${locationStr}) with ${participants} [${meeting.status}]`;
};

export const schedulingContextProvider: Provider = {
  name: "SCHEDULING_CONTEXT",
  description: "Provides context about upcoming meetings and scheduling requests",
  get: async (runtime: IAgentRuntime, message: Memory, _state: State) => {
    const schedulingService = runtime.getService<SchedulingService>("SCHEDULING");
    if (!schedulingService) {
      return { text: "" };
    }

    const entityId = message.entityId;
    if (!entityId) {
      return { text: "" };
    }

    const sections: string[] = [];

    try {
      const meetings = await schedulingService.getUpcomingMeetings(entityId);

      if (meetings.length > 0) {
        const proposedMeetings = meetings.filter((m) => m.status === "proposed");
        const confirmedMeetings = meetings.filter(
          (m) => m.status === "confirmed" || m.status === "scheduled"
        );

        if (proposedMeetings.length > 0) {
          sections.push("Meetings pending confirmation:");
          for (const meeting of proposedMeetings.slice(0, 3)) {
            sections.push(formatMeetingForContext(meeting, schedulingService));
          }
        }

        if (confirmedMeetings.length > 0) {
          sections.push("\nUpcoming confirmed meetings:");
          for (const meeting of confirmedMeetings.slice(0, 5)) {
            sections.push(formatMeetingForContext(meeting, schedulingService));
          }
        }
      }

      const availability = await schedulingService.getAvailability(entityId);
      if (availability) {
        const weeklyCount = availability.weekly.length;
        const exceptionsCount = availability.exceptions.length;

        sections.push(
          `\nUser has ${weeklyCount} recurring availability windows set (timezone: ${availability.timeZone})`
        );
        if (exceptionsCount > 0) {
          sections.push(`User has ${exceptionsCount} availability exceptions`);
        }
      } else {
        sections.push("\nUser has not set their availability yet");
      }
    } catch {
      // Silently fail if storage not ready
    }

    if (sections.length === 0) {
      return { text: "" };
    }

    return { text: `<scheduling_context>\n${sections.join("\n")}\n</scheduling_context>` };
  },
};

File: plugins/plugin-scheduling/rust/src/providers/scheduling_context.rs
//! SCHEDULING_CONTEXT provider.
//!
//! Provides the agent with context about upcoming meetings and scheduling requests.

use crate::service::SchedulingService;
use crate::types::{Meeting, MeetingStatus};

/// Format a meeting for display in context.
pub fn format_meeting_for_context(meeting: &Meeting, service: &SchedulingService) -> String {
    let time_str = service.format_slot(&meeting.slot);
    let participants: Vec<&str> = meeting.participants.iter().map(|p| p.name.as_str()).collect();
    let participants_str = participants.join(", ");

    let location_str = match meeting.location.location_type {
        crate::types::LocationType::InPerson => {
            format!("at {}", meeting.location.name.as_deref().unwrap_or("TBD"))
        }
        crate::types::LocationType::Virtual => "virtual meeting".to_string(),
        crate::types::LocationType::Phone => "phone call".to_string(),
    };

    format!(
        "- \"{}\" on {} ({}) with {} [{:?}]",
        meeting.title,
        time_str,
        location_str,
        participants_str,
        meeting.status
    )
}

/// SCHEDULING_CONTEXT provider.
pub struct SchedulingContextProvider;

impl SchedulingContextProvider {
    pub const NAME: &'static str = "SCHEDULING_CONTEXT";
    pub const DESCRIPTION: &'static str =
        "Provides context about upcoming meetings and scheduling requests";

    /// Get scheduling context for an entity.
    pub async fn get(
        service: &SchedulingService,
        entity_id: &str,
    ) -> String {
        let mut sections: Vec<String> = Vec::new();

        match service.get_upcoming_meetings(entity_id).await {
            Ok(meetings) if !meetings.is_empty() => {
                let proposed: Vec<&Meeting> = meetings
                    .iter()
                    .filter(|m| m.status == MeetingStatus::Proposed)
                    .collect();

                let confirmed: Vec<&Meeting> = meetings
                    .iter()
                    .filter(|m| {
                        m.status == MeetingStatus::Confirmed
                            || m.status == MeetingStatus::Scheduled
                    })
                    .collect();

                if !proposed.is_empty() {
                    sections.push("Meetings pending confirmation:".to_string());
                    for meeting in proposed.iter().take(3) {
                        sections.push(format_meeting_for_context(meeting, service));
                    }
                }

                if !confirmed.is_empty() {
                    sections.push("\nUpcoming confirmed meetings:".to_string());
                    for meeting in confirmed.iter().take(5) {
                        sections.push(format_meeting_for_context(meeting, service));
                    }
                }
            }
            _ => {}
        }

        match service.get_availability(entity_id).await {
            Ok(Some(availability)) => {
                let weekly_count = availability.weekly.len();
                let exceptions_count = availability.exceptions.len();
                sections.push(format!(
                    "\nUser has {} recurring availability windows set (timezone: {})",
                    weekly_count, availability.time_zone
                ));
                if exceptions_count > 0 {
                    sections.push(format!(
                        "User has {} availability exceptions",
                        exceptions_count
                    ));
                }
            }
            Ok(None) => {
                sections.push("\nUser has not set their availability yet".to_string());
            }
            Err(_) => {}
        }

        if sections.is_empty() {
            return String::new();
        }

        format!(
            "<scheduling_context>\n{}\n</scheduling_context>",
            sections.join("\n")
        )
    }
}
